```{r , echo=FALSE}
suppressPackageStartupMessages(library(ggplot2))
suppressPackageStartupMessages(library(seqinr))
```

# Meet magpie! ![An image of a magpie.](img/magpie-64x64.png)

## An on-demand R notebook ##

This interactive R notebook is running as a [Shiny][shiny] application
in R on a gizmo node that was snagged for a few days with grabcpu, to
emulate what having a real Shiny server would be like.

Hint: You can change things! Use `Ctrl-Alt-I` to insert an R code block,
use markdown syntax for text, and hit `F4` to recompile. Just be
careful, there's no save button yet. Reload the page if you do too
much damage.

## Loading data from silo

Since R is running on a gizmo node, we've got access to silo. This is
pretty neat, since that also means we have access to Galaxy output.

This is where our files live.

```{r }
database <- "/shared/silo_researcher/Matsen_F/MatsenGrp/data/galaxy-shapley/database"
```

Eventually this section this will be automatically generated and
inserted by Galaxy and magpie.

```{r }
## "[Genus] Grouped-by-specimen classification (CSV)",csv,84726a17127220c7
by.specimen <- read.csv(file.path(database, "/files/004/dataset_4880.dat"), header=TRUE, as.is=TRUE)

## "1.TCA.454Reads.1000.fasta",fasta,2a0d57490303b027
reads <- read.fasta(file.path(database, "files/001/dataset_1351.dat"))
```

Let's check if our data is there.

```{r }
head(by.specimen)
```

```{r }
str(reads[1])
```

Yay!

## Doing things

Now we can do pretty much whatever we want with Galaxy data (in any
format R can be made to understand). Like make pretty pictures using
the microbiome classification outputs we loaded earlier!

```{r , fig.height=8, fig.width=12}
library(ggplot2)
ggplot(subset(by.specimen, tax_name=="Lactobacillus")[1:20,], aes(x=specimen, y=placements)) + geom_point(aes(color=placements, size=tally))
```

I should note that the microbiome dataset loaded above and used for
this plot was generated by our Galaxy microbiome workflow when I was
testing it back in March. To be precise, it was generated Sun Mar 30
10:18:18 2014 (UTC). Here's how I know, and what Galaxy (and a user)
can see about a dataset:

![A screenshot of dataset metadata.](img/dataset-metadata.png)

Anyway, for tools like Martin and Dan's, I think this kind of approach
makes a lot more sense than the standard submit-and-wait Galaxy
experience, and I think is more in line with what they wanted. They
can write their analyses in R and markdown, have access to the
pipeline outputs, know where the data came from, and can:

1. Use Shiny's [gorgeous interface][gallery] to make dynamic,
user-friendly displays;
1. Generate a pretty ps or pdf, complete with timestamps and
provenance, binary versions, etc.; or
1. Show a page like this and have a little playground to tweak things
or poke at data.

And go click that link if you didn't. Seriously. At least look at
[this one][superzip].

## And more cool stuff!

In addition to presenting an editor, Shiny can also fetch text to
render from other sources:

* [runGist][rungist] -- Run a Shiny application from
https://gist.github.com
* [runGitHub][rungithub] -- Run a Shiny application from a GitHub
repository
* [runUrl][runurl] -- Run a Shiny application from a URL

There's also a standalone [Shiny server][server] capable of handling
more requests, available in both open-source and proprietary flavors,
and hosting available at [ShinyApps.io][hosted] (but that wouldn't
have access to Hutch resources).

[shiny]: http://shiny.rstudio.com/
[rungist]: http://shiny.rstudio.com/reference/shiny/latest/runGist.html
[rungithub]: http://shiny.rstudio.com/reference/shiny/latest/runGitHub.html
[runurl]: http://shiny.rstudio.com/reference/shiny/latest/runUrl.html
[server]: http://shiny.rstudio.com/deploy/
[hosted]: http://www.rstudio.com/shiny/hosted/
[gallery]: http://shiny.rstudio.com/gallery/
[superzip]: http://shiny.rstudio.com/gallery/superzip-example.html
